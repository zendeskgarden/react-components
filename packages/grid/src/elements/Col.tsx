/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { StyledCol } from '../styled';

export interface IColProps extends HTMLAttributes<HTMLDivElement> {
  /** Sizing for all breakpoints. */
  size?: number | string;
  /** Sizing for extra small breakpoints. */
  xs?: number | string | boolean;
  /** Sizing for small breakpoints. */
  sm?: number | string | boolean;
  /** Sizing for medium breakpoints. */
  md?: number | string | boolean;
  /** Sizing for large breakpoints. */
  lg?: number | string | boolean;
  /** Sizing for extra large breakpoints. */
  xl?: number | string | boolean;
  offsetXs?: number | string | boolean;
  offsetSm?: number | string | boolean;
  offsetMd?: number | string | boolean;
  offsetLg?: number | string | boolean;
  offsetXl?: number | string | boolean;
  /** Use flexbox alignment utilities to horizontally align */
  alignSelf?: 'start' | 'center' | 'end';
  justifyContent?: 'start' | 'center' | 'end' | 'around' | 'between';
  order?: any;
}

/**
 * Accepts all `<div>` attributes and events
 */
export const Col = React.forwardRef<HTMLDivElement, IColProps>((props, ref) => (
  <StyledCol ref={ref} {...props} />
));

Col.propTypes = {
  size: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  md: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  offsetXs: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  offsetSm: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  offsetMd: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  offsetLg: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  offsetXl: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  alignSelf: PropTypes.oneOf(['start', 'center', 'end']),
  justifyContent: PropTypes.oneOf(['start', 'center', 'end', 'around', 'between']),
  order: PropTypes.any
};
