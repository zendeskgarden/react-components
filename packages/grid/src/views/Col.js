/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import PropTypes from 'prop-types';
import styled from 'styled-components';
import classNames from 'classnames';
import { retrieveTheme } from '@zendeskgarden/react-theming';
import GridStyles from '@zendeskgarden/css-grid';

const COMPONENT_ID = 'grid.col';

const retrieveNumberedClass = (className, number, classes) => {
  if (typeof number === 'boolean') {
    classes.push(GridStyles[className]);
  } else if (typeof number !== 'undefined') {
    classes.push(GridStyles[`${className}-${number}`]);
  }
};

const retrieveColClassNames = ({
  size,
  xs,
  sm,
  md,
  lg,
  xl,
  offsetXs,
  offsetSm,
  offsetMd,
  offsetLg,
  offsetXl
} = {}) => {
  const output = [];

  retrieveNumberedClass('col', size, output);
  retrieveNumberedClass('col-xs', xs, output);
  retrieveNumberedClass('col-sm', sm, output);
  retrieveNumberedClass('col-md', md, output);
  retrieveNumberedClass('col-lg', lg, output);
  retrieveNumberedClass('col-xl', xl, output);
  retrieveNumberedClass('offset-xs', offsetXs, output);
  retrieveNumberedClass('offset-sm', offsetSm, output);
  retrieveNumberedClass('offset-md', offsetMd, output);
  retrieveNumberedClass('offset-lg', offsetLg, output);
  retrieveNumberedClass('offset-xl', offsetXl, output);

  if (output.length === 0) {
    output.push(GridStyles.col);
  }

  return output;
};

/**
 * Accepts all `<div>` props
 */
const Col = styled.div.attrs({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  className: props =>
    classNames(...retrieveColClassNames(props), {
      [GridStyles[`align-self-${props.alignSelf}`]]: props.alignSelf,
      [GridStyles[`justify-content-${props.justifyContent}`]]: props.justifyContent,
      [GridStyles[`order-${props.order}`]]: props.order
    })
})`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

Col.propTypes = {
  /** Sizing for all breakpoints. */
  size: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  /** Sizing for extra small breakpoints. */
  xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  /** Sizing for small breakpoints. */
  sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  /** Sizing for medium breakpoints. */
  md: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  /** Sizing for large breakpoints. */
  lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  /** Sizing for extra large breakpoints. */
  xl: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  offsetXs: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  offsetSm: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  offsetMd: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  offsetLg: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  offsetXl: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
  /** Use flexbox alignment utilities to horizontally align */
  alignSelf: PropTypes.oneOf(['start', 'center', 'end']),
  justifyContent: PropTypes.oneOf(['start', 'center', 'end', 'around', 'between']),
  order: PropTypes.any
};

/** @component */
export default Col;
