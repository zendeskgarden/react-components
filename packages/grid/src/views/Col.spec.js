/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { shallow } from 'enzyme';
import Col from './Col';

describe('Col', () => {
  it('renders default styling', () => {
    const wrapper = shallow(<Col />);

    expect(wrapper).toHaveClassName('col');
  });

  describe('Sizing', () => {
    const sizes = [
      {
        size: 4
      },
      {
        sm: true
      },
      {
        xs: 6
      },
      {
        sm: 6
      },
      {
        md: 6
      },
      {
        lg: 6
      },
      {
        xl: 6
      }
    ];

    sizes.forEach(props => {
      const key = Object.keys(props)[0];

      it(`renders ${key} if provided`, () => {
        const wrapper = shallow(<Col {...props} />);
        let className;

        if (key === 'size') {
          className = `col-${props[key]}`;
        } else if (typeof props[key] === 'number') {
          className = `col-${key}-${props[key]}`;
        } else {
          className = `col-${key}`;
        }

        expect(wrapper).toHaveClassName(className);
      });
    });
  });

  describe('Offsets', () => {
    ['offsetXs', 'offsetSm', 'offsetMd', 'offsetLg', 'offsetXl'].forEach(offset => {
      const props = {
        [offset]: 6
      };

      const classes = {
        offsetXs: 'offset-xs',
        offsetSm: 'offset-sm',
        offsetMd: 'offset-md',
        offsetLg: 'offset-lg',
        offsetXl: 'offset-xl'
      };

      it(`renders ${offset} if provided`, () => {
        const wrapper = shallow(<Col {...props} />);

        expect(wrapper).toHaveClassName(`${classes[offset]}-${props[offset]}`);
      });
    });
  });

  describe('Align Self', () => {
    ['start', 'center', 'end'].forEach(alignment => {
      it(`renders ${alignment} self alignment if provided`, () => {
        const wrapper = shallow(<Col alignSelf={alignment} />);

        expect(wrapper).toHaveClassName(`align-self-${alignment}`);
      });
    });
  });

  describe('Justify Content', () => {
    ['start', 'center', 'end', 'around', 'between'].forEach(justifyContent => {
      it(`renders ${justifyContent} justify content if provided`, () => {
        const wrapper = shallow(<Col justifyContent={justifyContent} />);

        expect(wrapper).toHaveClassName(`justify-content-${justifyContent}`);
      });
    });
  });

  describe('Order', () => {
    it('renders pseudo order if provided', () => {
      const wrapper = shallow(<Col order="first" />);

      expect(wrapper).toHaveClassName('order-first');
    });

    it('renders string based order if provided', () => {
      const wrapper = shallow(<Col order="md-12" />);

      expect(wrapper).toHaveClassName('order-md-12');
    });
  });
});
