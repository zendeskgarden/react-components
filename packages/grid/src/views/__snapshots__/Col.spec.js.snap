// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col Align Self renders center self alignment if provided 1`] = `
<div
  className="col align-self-center "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Align Self renders end self alignment if provided 1`] = `
<div
  className="col align-self-end "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Align Self renders start self alignment if provided 1`] = `
<div
  className="col align-self-start "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Justify Content renders around justify content if provided 1`] = `
<div
  className="col justify-content-around "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Justify Content renders between justify content if provided 1`] = `
<div
  className="col justify-content-between "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Justify Content renders center justify content if provided 1`] = `
<div
  className="col justify-content-center "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Justify Content renders end justify content if provided 1`] = `
<div
  className="col justify-content-end "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Justify Content renders start justify content if provided 1`] = `
<div
  className="col justify-content-start "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Offsets renders offsetLg if provided 1`] = `
<div
  className="offset-lg-6 "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Offsets renders offsetMd if provided 1`] = `
<div
  className="offset-md-6 "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Offsets renders offsetSm if provided 1`] = `
<div
  className="offset-sm-6 "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Offsets renders offsetXl if provided 1`] = `
<div
  className="offset-xl-6 "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Offsets renders offsetXs if provided 1`] = `
<div
  className="offset-xs-6 "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Order renders pseudo order if provided 1`] = `
<div
  className="col order-first "
  data-garden-id="grid.col"
  data-garden-version="version"
  order="first"
/>
`;

exports[`Col Order renders string based order if provided 1`] = `
<div
  className="col order-md-12 "
  data-garden-id="grid.col"
  data-garden-version="version"
  order="md-12"
/>
`;

exports[`Col Sizing renders lg if provided 1`] = `
<div
  className="col-lg-6 "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Sizing renders md if provided 1`] = `
<div
  className="col-md-6 "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Sizing renders size if provided 1`] = `
<div
  className="col-4 "
  data-garden-id="grid.col"
  data-garden-version="version"
  size={4}
/>
`;

exports[`Col Sizing renders sm if provided 1`] = `
<div
  className="col-sm "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Sizing renders sm if provided 2`] = `
<div
  className="col-sm-6 "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Sizing renders xl if provided 1`] = `
<div
  className="col-xl-6 "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col Sizing renders xs if provided 1`] = `
<div
  className="col-xs-6 "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;

exports[`Col renders default styling 1`] = `
<div
  className="col "
  data-garden-id="grid.col"
  data-garden-version="version"
/>
`;
